### PR TITLE
 feat(testcafé) : file Sharing scenario

### PR DESCRIPTION
**Pull request by CozyKourai**
* One commit with all the new data-test-id

* One commit with new tests and data (Check trello card for expected steps)

* For now, both unauthenticated tests are skipped, because they always failed because of a drive error.
cf. https://mattermost.cozycloud.cc/test-team/pl/xx9xfxrd3fn65byhtfuhnnw3ww
the same error poped-up for the folder sharing scenario and was fix in https://github.com/cozy/cozy-drive/pull/1420. We need the same kind of fix for files

### CODE REVIEW

**Comment by CozyKourai** [comment]
[(S) SCN - partage par lien d'un fichier](https://trello.com/c/mQzLJmjK/1597-s-scn-partage-par-lien-dun-fichier)

File: testcafe/tests/drive/drive_nav.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import { Selector, Role } from 'testcafe'
```
Comment (Crash--) [inline testcafe/tests/drive/drive_nav.js]: Selector & Role not used

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import { Selector, Role } from 'testcafe'
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: Selector not used

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import { driveUser } from '../helpers/roles'
+import {
+  TESTCAFE_DRIVE_URL,
+  isExistingAndVisibile,
+  getCurrentDateTime,
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: not used

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import { Selector, Role } from 'testcafe'
+import { driveUser } from '../helpers/roles'
+import {
+  TESTCAFE_DRIVE_URL,
+  isExistingAndVisibile,
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: not used

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  await publicDrivePage.checkActionMenuPublicMobile()
+  await t
+    .setNativeDialogHandler(() => true)
+    .click(publicDrivePage.btnPublicMobileDownload)
+    .click(publicDrivePage.btnPublicMoreMenu) //need to re-open the more menu
+    .click(publicDrivePage.btnPublicMobileCreateCozy)
+  await publicDrivePage.checkCreateCozy()
+
+  await t.resizeWindow(1280, 1024) //Back to desktop
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: to change

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+//************************
+fixture.skip`Drive : Access a folder public link`.page`${TESTCAFE_DRIVE_URL}/`
+  .beforeEach(async t => {
+    await t.useRole(Role.anonymous())
+  })
+  .afterEach(async t => {
+    await publicDrivePage.checkLocalFile('files.zip')
+    await publicDrivePage.deleteLocalFile('files.zip')
+  })
+test('Drive : Access a file public link (desktop)', async t => {
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: Access a file? It's more about `download` here, no?

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  })
+  await t.navigateTo(data.sharingLink)
+  await publicDrivePage.waitForViewer()
+
+  await publicDrivePage.checkActionMenuPublicMobile()
+  await t
+    .setNativeDialogHandler(() => true)
+    .click(publicDrivePage.btnPublicMobileDownload)
+    .click(publicDrivePage.btnPublicMoreMenu) //need to re-open the more menu
+    .click(publicDrivePage.btnPublicMobileCreateCozy)
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: So we have a diff between mobile & not mobile ?

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
+import { Selector, Role } from 'testcafe'
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: Selector

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+import { driveUser } from '../helpers/roles'
+import {
+  TESTCAFE_DRIVE_URL,
+  isExistingAndVisibile,
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: to be removed

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  getCurrentDateTime,
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: to be removed

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+import Data from '../helpers/data'
+const data = new Data()
+
+import DrivePage from '../pages/drive-model'
+const drivePage = new DrivePage()
+
+import PublicDrivePage from '../pages/drive-model-public'
+const publicDrivePage = new PublicDrivePage()
+
+const file = '[cozy]QA_table_ APPS.pdf'
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: We should have a constant file with this kind of variables.

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+//************************
+fixture`Drive : Access a folder public link`.page`${TESTCAFE_DRIVE_URL}/`
+  .beforeEach(async t => {
+    await t.useRole(Role.anonymous())
+  })
+  .afterEach(async t => {
+    await publicDrivePage.checkLocalFile('files.zip')
+    await publicDrivePage.deleteLocalFile('files.zip')
+  })
+test('Drive : Access a folder public link (desktop)', async t => {
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: idem it's more about downloading than accessing, no? We don't check if something specific to this file is displayed on the screen, right?

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+  await publicDrivePage.waitForLoading()
+  await publicDrivePage.checkActionMenuPublicMobile()
+  await t
+    .setNativeDialogHandler(() => true)
+    .click(publicDrivePage.btnPublicMobileDownload)
+    .click(publicDrivePage.btnPublicMoreMenu) //need to re-open the more menu
+    .click(publicDrivePage.btnPublicMobileCreateCozy)
+  await publicDrivePage.checkCreateCozy()
+
+  await t.resizeWindow(1280, 1024) //Back to desktop
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: to be edited

File: testcafe/tests/helpers/data.js [hunk 14]
```diff
@@ -0,0 +1,5 @@
+export default class Data {
+  constructor() {
+    this.sharingLink = ''
+  }
+}
```
Comment (Crash--) [inline testcafe/tests/helpers/data.js]: why ?

File: testcafe/tests/helpers/utils.js [hunk 15]
```diff
@@ -36,3 +38,22 @@ export async function isExistingAndVisibile(selector, selectorName) {
     .ok(`'${selectorName}' is not visible`)
   console.log(`'${selectorName}' exists and is visible!`)
 }
+
+export function getCurrentDateTime() {
```
Comment (Crash--) [inline testcafe/tests/helpers/utils.js]: no need to export

File: testcafe/tests/drive/folder_sharing_scenario.js (hunk 12 above)
Comment (CozyKourai) [reply testcafe/tests/drive/folder_sharing_scenario.js]: We don't check something specific to the file, just that there are buttons to download it, and to create a new cozy.

File: testcafe/tests/drive/file_sharing_scenario.js (hunk 7 above)
Comment (CozyKourai) [reply testcafe/tests/drive/file_sharing_scenario.js]: Yes, in mobile view, the download button and create cozy button are in a [...] menu

File: testcafe/tests/drive/folder_sharing_scenario.js (hunk 12 above)
Comment (Crash--) [reply testcafe/tests/drive/folder_sharing_scenario.js]: So I would prefer to have: 

test('Download a folder from a public link') 

test('Create a cozy from a public link')

File: testcafe/tests/drive/folder_sharing_scenario.js (hunk 11 above)
Comment (CozyKourai) [reply testcafe/tests/drive/folder_sharing_scenario.js]: Added in data.js
(and I removed the constructor)

File: testcafe/tests/drive/drive_nav.js [hunk 16]
```diff
@@ -11,7 +11,7 @@ fixture`DRIVE - NAV`.page`${TESTCAFE_DRIVE_URL}/`.beforeEach(async t => {
 
 test('Drive Navigation Desktop Resolution: Drive, Recent, Sharing, Trash', async () => {
   //Check Menu and links. Go to page. Check main menu on each page
-  await isExistingAndVisibile(page.sidebar, 'Sidebar')
+  isExistingAndVisibile(page.sidebar, 'Sidebar')
```
Comment (Crash--) [inline testcafe/tests/drive/drive_nav.js]: await ?

File: testcafe/tests/drive/folder_sharing_scenario.js [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+//************************
+// Public (no authentification)
+//************************
+fixture`Drive : Access a folder public link`.page`${TESTCAFE_DRIVE_URL}/`
+  .beforeEach(async t => {
+    await t.useRole(Role.anonymous())
+    await setDownloadPath(`${homedir()}/Downloads`)
+  })
+  .afterEach(async () => {
+    const filepath = `${homedir()}/Downloads/files.zip`
```
Comment (Crash--) [inline testcafe/tests/drive/folder_sharing_scenario.js]: we should have this `/Download/files.zip` in a constant somewhere with the PDF file etc.

File: testcafe/tests/drive_sharing.js [hunk 18]
```diff
@@ -5,20 +5,15 @@ import {
...
 } from './helpers/utils'
-import { TESTCAFE_DRIVE_URL, FOLDER_DATE_TIME } from './helpers/utils'
+import { TESTCAFE_DRIVE_URL, FOLDER_DATE_TIME } from '../helpers/utils'
 const homedir = require('os').homedir
-
-import Data from './helpers/data'
-const data = new Data()
-
-import DrivePage from './pages/drive-model'
+let data = require('../helpers/data')
```
Comment (Crash--) [inline testcafe/tests/drive_sharing.js]: import { filePDF }  from '../helpers/data'

File: testcafe/tests/drive_sharing.js [hunk 19]
```diff
@@ -39,7 +34,7 @@ test('Drive : from Drive, go in a folder, upload a file, and share it', async t
   await drivePage.openActionMenu()
   await t.pressKey('esc') //close action Menu
 
-  await drivePage.uploadFiles([`../data/${file}`])
+  await drivePage.uploadFiles([`../data/${data.filePDF}`])
```
Comment (Crash--) [inline testcafe/tests/drive_sharing.js]: We should keep the full path into the variable, WDYT?

File: testcafe/tests/drive/file_sharing_scenario.js [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+import { driveUser } from '../helpers/roles'
+import {
+  deleteLocalFile,
+  checkLocalFile,
+  setDownloadPath,
+  TESTCAFE_DRIVE_URL,
+  FOLDER_DATE_TIME
+} from '../helpers/utils'
+const homedir = require('os').homedir
+let data = require('../helpers/data')
```
Comment (Crash--) [inline testcafe/tests/drive/file_sharing_scenario.js]: import { var1, var2 } from '../'

File: testcafe/tests/photos/photos_crud.js [hunk 21]
```diff
@@ -12,14 +12,14 @@ fixture`PHOTOS - CRUD`.page`${TESTCAFE_PHOTOS_URL}/`.beforeEach(async t => {
...
 })
 
 test('Uploading 3 pcis from Photos view', async t => {
   await page.uploadPhotos([
-    '../data/IMG-JPG.jpg',
-    '../data/IMG-PNG.png',
-    '../data/IMG-GIF.gif'
+    '../../data/IMG-JPG.jpg',
+    '../../data/IMG-PNG.png',
+    '../../data/IMG-GIF.gif'
```
Comment (Crash--) [inline testcafe/tests/photos/photos_crud.js]: We should put these to a constante too ;)

File: testcafe/tests/drive/folder_sharing_scenario.js (hunk 12 above)
Comment (CozyKourai) [reply testcafe/tests/drive/folder_sharing_scenario.js]: new name : *Drive : Access a folder public link, download the file(s), and check the 'create Cozy' link`*

File: testcafe/tests/drive_sharing.js (hunk 19 above)
Comment (CozyKourai) [reply testcafe/tests/drive_sharing.js]: I added 2 vars : one for file name, and one for the data folder path, as I re-used the file name later (and data folder path is always the same anyway!)

File: testcafe/tests/pages/drive-model-public.js [hunk 22]
```diff
@@ -62,27 +86,50 @@ export default class PublicDrivePage {
...
+    } else if (type == 'folder') {
+      await t
+        .expect(this.logo.exists)
+        .notOk('Logo exists')
+        .expect(this.toolbarFolderPublic.exists)
+        .notOk('toolbar_file exists')
+        .expect(this.btnPublicCreateCozyFolder.exists)
+        .notOk('Create Cozy button (desktop) exists')
+      await isExistingAndVisibile(this.btnPublicMoreMenuFolder, '[...] Button')
+      await t.click(this.btnPublicMoreMenuFolder)
```
Comment (Crash--) [inline testcafe/tests/pages/drive-model-public.js]: Are you sure you need to split it that way? It seems that except the first expect the rest is the same?

File: testcafe/tests/pages/drive-model-public.js [hunk 23]
```diff
@@ -62,27 +86,50 @@ export default class PublicDrivePage {
       .expect(this.btnPublicMobileDownload.exists)
       .notOk('Create my Cozy Button (mobile) exists')
     await isExistingAndVisibile(this.logo, 'Logo')
-    await isExistingAndVisibile(this.toolbarFilesPublic, 'toolbarFiles')
+    if (type == 'file') {
```
Comment (Crash--) [inline testcafe/tests/pages/drive-model-public.js]: ===

File: testcafe/tests/pages/drive-model-public.js [hunk 24]
```diff
@@ -62,27 +86,50 @@ export default class PublicDrivePage {
...
+        this.btnPublicCreateCozyFile,
+        'Create my Cozy Button'
+      )
+    } else if (type == 'folder') {
+      await isExistingAndVisibile(this.toolbarFolderPublic, 'toolbarFolder')
+      await isExistingAndVisibile(
+        this.btnPublicCreateCozyFolder,
+        'Create my Cozy Button'
+      )
+    }
```
Comment (Crash--) [inline testcafe/tests/pages/drive-model-public.js]:
```js
const isFile = type === 'file' ? true : false
await isExistingAndVisibile(isFile ? this.toolbarFilesPublic : toolbarFolderPublic , 'toolbarFiles')
      await isExistingAndVisibile(
        isFile ? this.btnPublicCreateCozyFile : this.btnPublicCreateCozyFolder,
        'Create my Cozy Button'
      )
```

When you create a if / else block we're excepting that you do very distinct stuff. This is not the case here, you just change a selector so an inline if is more readable?

File: testcafe/tests/pages/drive-model-public.js (hunk 22 above)
Comment (Crash--) [reply testcafe/tests/pages/drive-model-public.js]: see https://github.com/cozy/cozy-drive/pull/1433/files#r257228422

**Comment by CozyKourai** [comment]
[(𝐒) SCN - partage par lien d'un fichier](https://trello.com/c/mQzLJmjK/1597-%F0%9D%90%92-scn-partage-par-lien-dun-fichier)

**Comment by CozyKourai** [comment]
[(𝐒) SCN - partage par lien d'un album](https://trello.com/c/93PI94N2/1601-%F0%9D%90%92-scn-partage-par-lien-dun-album)

**Pull request merged**